src/set_theory/ordinal/arithmetic.lean

```diff
@@ -2376,4 +2390,15 @@ lemma rank_strict_anti [preorder α] [well_founded_gt α] :
   strict_anti (rank $ @is_well_founded.wf α (>) _) :=
 λ _ _, rank_lt_of_rel $ @is_well_founded.wf α (>) _
 
+@[simp] lemma rank_eq_typein [is_well_order α r] :
+  well_founded.rank (@is_well_founded.wf α r _) a = ordinal.typein r a :=
```

Comment:
Wait, aren't `well_founded.rank` and `ordinal.typein` quite literally the same thing? Why aren't they defined one in term of the other?

Reply:
Good point. This would be a large refactor, since a lot of proofs would be need to moved around. `ordinal.typein` can be defined almost immediately, while `ordinal.rank` needs `ordinal.sup`. I'll look into the viability of this.

Reply:
That said. We might also want to define a version of `well_founded.rank` for an entire well-founded relation, analogous to `ordinal.rank`. In that case there would be no hope of a def-eq, since `ordinal.rank` is def-eq to the type constructor of `ordinal` itself. So maybe it's better and simpler to keep these separate and just prove the equalities.

Reply:
I don't understand your comment. `ordinal.typein` and `well_founded.rank` can each be defined in terms of the other (`is_well_order α r` and `well_founded r` are just a typeclass bundling ). So pick whichever is easiest and define the other in terms of it (if we even need both versions, really).

Reply:
`typein` is specifically for well-orders, while `rank` is for well-founded relations. The latter is more general.

Reply:
Oh right, I missed this. Then we should redefine `typein` in terms of `rank`.

Reply:
Yeah, that's the suggestion my comment addresses.